### PR TITLE
[META-46]: Split project into Gateway service and Graphql server service

### PR DESCRIPTION
**Pull request by pczern**
*Pending/Waiting for Env variables to be added to GitHub by Project Leads so tests can pass*

### Story
graphql client libraries that access multiple graphql endpoints are a hacky solution and generally not natively supported by the current clients. Thus the full schema including all services `Supabase` and our `Graphql Server` should be provided by the backend. To accomplish this, there is something called GraphQL Federation, that will combine schemas from different GraphQL servers. We have used Mercurius to add GraphQL functionality to our Fastify server. Federation with Mercurius as a Gateway was not possible as it requires the Apollo Federation v1 spec. This spec requires additional fields on the GraphQL schema provided by the GraphQL server. The services we use don't expose these fields, making it impossible to use this kind of Federation. Note it already exists an Apollo Federation v2 spec, which also requires special fields.
Therefore a solution had to be found how to connect to multiple GraphQL sources without having to rely on the Apollo Federation spec. An alternative to Federation is schema stitching, were schemas are manually combined with tools.
[GraphQL Mesh](https://the-guild.dev/graphql/mesh/docs)  gives us the best of both worlds, an open source service allowing us to setup a gateway to access remote GraphQL servers, add authentication capablities, GraphQL transforms/renames, Caching and more.

### Description
Split Backend into two services called `graphql_gateway` and `graphql_server` via [NPM workspaces](https://docs.npmjs.com/cli/v9/using-npm/workspaces). Implement GraphQL Gateway service with GraphQL Mesh for Supabase and `graphql_server` service. Gateway will verify for a set `jwt` cookie, to allow access to data.

### How I implemented
* Split project with NPM workspaces into `./packages/graphql_gateway` & `./packages/graphql_server`
-> Three package.json (top level and 2 services)
* `graphql_server` service is the server we already have
* `graphql_gateway` is the new gateway service
* Move ESLint, Prettier, Husky to top level
* Add `Dockerfile`s to `./packages` and setup top level `docker-compose.yml`
* Add/Adjust .dockerignore, .prettierignore, .gitignore
* Move 2 `tsconfig.json` respectivly to `./packages`
* Load env variables via `env.sh` script in `./packages/**/package.json`'s from `.env.dev`
This is accomplished by sourcing them in the `npm run` script (the script is referenced from the config key in `package.json`
* Setup GraphQL Mesh Gateway with Supabase and `graphql_server` service as source
* Add Authentication via set `jwt` cookie to GraphQL Mesh via envelop plugin system using [useGenericAuth](https://the-guild.dev/graphql/envelop/plugins/use-generic-auth)
* Simply npm run schema for all services, and toplevel access
* Update Readme's with simpler npm run schema
### Other
* As `npm run prepare` is run during build with Dockerfile, prevent it from running when `NODE_ENV=production`
* ScrapeRepos.cjs had to run Prettier (unchanged functionality)

### CODE REVIEW

File: README.md
Comment (cnk1dev) [inline README.md:6]: Add info on which env vars need to be set and where to find their values

File: README.md
Comment (cnk1dev) [inline README.md:33]: I would personally remove the graphql_ prefix from the packages everywhere

**Review by cnk1dev — CHANGES_REQUESTED** [review]
Looks good in terms of code quality, I can't really say much about the implementation itself as I don't have the necessary backend development experience, especially with GraphQL.

File: packages/graphql_server/src/server.ts
Comment (pczern) [reply packages/graphql_server/src/server.ts:6]: resolved

File: packages/graphql_server/tsconfig.json
Comment (pczern) [reply packages/graphql_server/tsconfig.json:1]: will be left for now, so everybody knows what's going on

File: README.md
Comment (pczern) [reply README.md:6]: resolved

File: README.md
Comment (pczern) [reply README.md:33]: agreed to keep

File: packages/graphql_gateway/README.md
Comment (pczern) [reply packages/graphql_gateway/README.md]: resolved

**Comment by pczern** [comment]
@cnk1dev comments resolved
@thomas-woehrle and @cnk1dev waiting for approval

**Review by cnk1dev — APPROVED** [review]

**Review by thomas-woehrle — COMMENTED** [review]
### RESOLVED
Issue with docker file

File: packages/graphql_gateway/README.md
Comment (pczern) [reply packages/graphql_gateway/README.md:19]: once the frontend has the jwt token, it will already be verified with Supabase

File: packages/graphql_gateway/src/envelopPlugins.ts
Comment (thomas-woehrle) [reply packages/graphql_gateway/src/envelopPlugins.ts:6]: Restarting VScode fixed it

**Review by thomas-woehrle — APPROVED** [review]

**Pull request merged**
